runtime/vm/jfr.cpp

```diff
@@ -720,6 +738,104 @@ jfrExecutionSampleCallback(J9VMThread *currentThread, IDATA handlerKey, void *us
+
+		omrthread_process_time_t currentProcCPUTime = {0};
+		intptr_t rc = omrthread_get_process_times(&currentProcCPUTime);
+
+		UDATA success = 0;
+		U_64 currentTime = (U_64)j9time_current_time_nanos(&success);
+		if (rc < 0 || 0 == success) {
+			jfrEvent->jvmUser = -1;
+			jfrEvent->jvmSystem = -1;
+		} else {
+			omrthread_process_time_t *prevProcCPUTime = &currentThread->javaVM->jfrState.prevProcCPUTime;
+			I_64 *prevProcTimestamp = &currentThread->javaVM->jfrState.prevProcTimestamp;
+			if (-1 == *prevProcTimestamp) {
+				jfrEvent->jvmUser = 0;
+				jfrEvent->jvmSystem = 0;
+			} else {
+				U_64 timeDelta = currentTime - *prevProcTimestamp;
+				jfrEvent->jvmUser = OMR_MIN((currentProcCPUTime._userTime - prevProcCPUTime->_userTime) / ((double)numberOfCpus * timeDelta), 1.0);
+				jfrEvent->jvmSystem = OMR_MIN((currentProcCPUTime._systemTime - prevProcCPUTime->_systemTime) / ((double)numberOfCpus * timeDelta), 1.0);
+			}
+			*prevProcCPUTime = currentProcCPUTime;
+			*prevProcTimestamp = currentTime;
+		}
+
+		J9SysinfoCPUTime currentSysCPUTime = {0};
+		rc = omrsysinfo_get_CPU_utilization(&currentSysCPUTime);
+		if (rc < 0) {
+			jfrEvent->machineTotal = rc;
+		} else {
+			J9SysinfoCPUTime *prevSysCPUTime = &currentThread->javaVM->jfrState.prevSysCPUTime;
+			if (-1 == prevSysCPUTime->timestamp) {
+				jfrEvent->machineTotal = 0;
+			} else {
+				jfrEvent->machineTotal = OMR_MIN((currentSysCPUTime.cpuTime - prevSysCPUTime->cpuTime) / ((double)numberOfCpus * (currentSysCPUTime.timestamp - prevSysCPUTime->timestamp)), 1.0);
+			}
+			*prevSysCPUTime = currentSysCPUTime;
+		}
+	}
+}
+
+void
+jfrThreadCPULoad(J9VMThread *currentThread, J9VMThread *sampleThread)
+{
+	PORT_ACCESS_FROM_VMC(currentThread);
+
+	J9JFRThreadCPULoad *jfrEvent = (J9JFRThreadCPULoad*)reserveBuffer(currentThread, sizeof(*jfrEvent));
+	if (NULL != jfrEvent) {
+		initializeEventFields(currentThread, (J9JFREvent *)jfrEvent, J9JFR_EVENT_TYPE_THREAD_CPU_LOAD);
+		/* Currently omrthread_get_user_time is not supported on all platforms.
+		 * If not supported, system time will be the total CPU time.
+		 */
+		I_64 cpuTime = (I_64)omrthread_get_cpu_time(sampleThread->osThread);
+		I_64 userTime = (I_64)omrthread_get_user_time(sampleThread->osThread);
+		UDATA success = 0;
+		U_64 currentTime = (U_64)j9time_current_time_nanos(&success);
+		if (0 == success || -1 == cpuTime) {
+			jfrEvent->user = -1;
+			jfrEvent->system = -1;
+		} else {
+			ThreadJFRState* jfrState = &sampleThread->jfrState;
+			I_64 *prevTimestamp = &jfrState->prevTimestamp;
+			if (-1 == *prevTimestamp) {
+				jfrEvent->user = 0;
+				jfrEvent->system = 0;
+			} else {
+				U_64 timeDelta = currentTime - *prevTimestamp;
+				UDATA numberOfCpus = j9sysinfo_get_number_CPUs_by_type(J9PORT_CPU_ONLINE);
+				if (-1 == userTime) {
+					jfrEvent->user = -1;
+					jfrEvent->system = OMR_MIN((cpuTime - jfrState->prevSysTime) / ((double)numberOfCpus * timeDelta), 1.0);
```

Comment:
do you need to divide by number of CPUs? a thread can only be on one CPU at a time

Reply:
I think you're correct. I didn't think about this when copying from the CPULoad part.